controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1173,6 +1180,67 @@ private long findSegmentSplitsMerges(List<StreamSegmentRecord> referenceSegments
         });
     }
 
+    @Override
+    public CompletableFuture<Void> recordCommitOffsets(final UUID txnId, final Map<Long, Long> commitOffsets) {
+        // The transaction may already have been committed and its record removed and this could be an idempotent run. 
+        // So this method will ignore any data not found exceptions that are thrown while trying to retrieve active 
+        // transaction record. 
+        // This method updates the commit offsets if they are empty. Otherwise it ignores them. 
+        int epoch = RecordHelper.getTransactionEpoch(txnId);
+        return Futures.exceptionallyExpecting(getActiveTx(epoch, txnId)
+                .thenCompose(txnRecord -> {
+                    ActiveTxnRecord activeTxnRecord = txnRecord.getObject();
+                    Preconditions.checkArgument(activeTxnRecord.getTxnStatus().equals(TxnStatus.COMMITTING));
+                    if (activeTxnRecord.getCommitOffsets().isEmpty()) {
+                        ActiveTxnRecord updated = new ActiveTxnRecord(activeTxnRecord.getTxCreationTimestamp(),
+                                activeTxnRecord.getLeaseExpiryTime(),
+                                activeTxnRecord.getMaxExecutionExpiryTime(),
+                                TxnStatus.COMMITTING,
+                                activeTxnRecord.getWriterId(),
+                                activeTxnRecord.getCommitTime(),
+                                activeTxnRecord.getCommitOrder(),
+                                ImmutableMap.copyOf(commitOffsets));
+                        return Futures.toVoid(updateActiveTx(epoch, txnId, new VersionedMetadata<>(updated, txnRecord.getVersion())));
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                }), DATA_NOT_FOUND_PREDICATE, null);
+    }
+
+    /**
+     * This method takes the list of transactions in the committing transactions record, and tries to report marks
+     * for writers for these transactions, if they exist. 
+     * WriterId, commit time and commit offsets are recorded in ActiveTxnRecord for each transaction. 
+     * For transactions where above fields are present, a mark is recorded for them. 
+     * This method will ignore any INVALID_TIME or INVALID_POSITION related failures in noting marks for writers.
+     * This is because those will typically arise from idempotent commit case where this and a transaction with higher 
```

Comment:
Say we have 3 txns from writer1 in following order txn1 (time = t1), txn2 (time = t2), txn3 (time = t3). 
We batch them together and commit in order :
1. commit txn1 at pos1
2. note marker for pos1, 
3. commit txn2 at pos2
4. note marker for pos2

Say processing failed here and the whole commit workflow is retried. 

So writer's latest noted position is pos2 and time t2. 

now if commit workflow is tried again, txn1 commit will have pos1 which is before pos2 and noteWriterTime will fail with INVALID_TIME/INVALID_POSITION.
This will be ignored and we continue with txn2. This will also fail with INVALID_TIME. 

Then we proceed to txn3, we commit it at pos3. The writermark is updated with pos3/t3.

Reply:
Maybe you can suggest a better phrasing for the above.